### ISSUE TITLE
Reticulate does not capture PDF in IPython rich outputs

### ISSUE THREAD

**Issue by mwouts** [rstudio/reticulate#281]
I sometime use IPython's rich outputs to display PDF files (my use case is : pdf versions of plotly graphs generated in python, cf. https://github.com/mwouts/plotly_offline_export). 
Unfortunately these are not captured by reticulate. Is this something that could be implemented ?

Attached is an example of such a PDF rich output. 
[reticulate_pdf.Rnw](https://github.com/rstudio/reticulate/files/2048271/reticulate_pdf_Rnw.txt)

**Comment by jjallaire** [comment]
It's probably possible but might be fairly expensive to implement. Could you provide a link to the documentation on iPython rich outputs? Does this also cover iPython widgets (e.g. plotly)

**Comment by mwouts** [comment]
Hello @jjallaire, thanks for your answer. Unfortunately I am not aware of a complete documentation on ipython rich outputs. I've mostly learnt from trial and error. This [demo](https://github.com/ipython/ipython/blob/master/examples/IPython%20Kernel/Rich%20Output.ipynb), explains well the opposite direction (how to create rich outputs). 

My experience is that
1. One can intercept the rich output with IPython.utils.capture, as at https://github.com/mwouts/plotly_offline_export/blob/master/plotly_offline_export.py#L144-L150. Indeed rich outputs are sometime HTML, or javascript (but I am not sure that a direct inclusion in knitr's output would work). In the case of plotly graphs, I've practiced rich outputs to *enrich* the javascript with plain images (for emails) or pdfs (for beamers).
2. Jupyter kernels do return rich output objects. In the pre-reticulate area I used them to execute python chunks with knitr (cf. [this pull request](https://github.com/yihui/knitr/pull/1424), and the [standalone package](https://github.com/mwouts/ipython_from_R)). I was able to save png and pdf outputs to local files for inclusion in knitr, cf. for instance https://github.com/mwouts/ipython_from_R/blob/master/inst/python/ipython_exec.py#L270-L310.

For inclusion in reticulate, I think IPython.utils.capture would be the natural way to go. I'll work on a example on how to capture a matplotlib plot and follow-up on this conversation.

**Comment by mwouts** [comment]
The promised example for capturing rich outputs is [here](https://github.com/mwouts/notebooks/blob/master/Capturing%20ipython%20rich%20output.ipynb). I've documented PNG only, but in my experience it works for any type of output (json, javascript, html, pdf, text...).

The example works well in jupyter notebooks, but... I just tested in a plain python session, and the same script does not capture anything there. I guess you need it to work with plain python, right? Would you like me to ask the question at https://github.com/ipython/ipython ?

**Comment by jjallaire** [comment]
Yes, we would need to be able to capture the output in a plain python session (sort of like we do in R with `knit_print` and html dependencies). If you could ask that question there that would be much appreciated!

**Comment by mwouts** [comment]
Sure. I've opened https://github.com/ipython/ipython/issues/11163, with a simple example that works in jupyter, and not in python.

**Comment by mwouts** [comment]
I got an answer for the ipython question. Capture of rich outputs requires the instanciation of a `InteractiveShell` class with
```
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell().instance()
```

I've [tested](https://github.com/mwouts/notebooks/blob/master/reticulate%20and%20capture.Rmd) the capture on
* png images
* plotly javascript library
* plotly json (data) and javascript plots

I think it should actually work on every output displayed with IPython's display method.

I do not know yet how to capture mpl plots with this technology - even worse, reticulate seems to hang on them when I instanciate the `InteractiveShell`.

**Comment by jjallaire** [comment]
That's great news!

What are mpl plots?

**Comment by mwouts** [comment]
Yes, indeed ! I think we should further experiment (and test in a broad range of python/R/os combinations). Also, we need to pay attention to side effects - the first one I've seen for now is the incompatibility with matplotlib (mpl) plots...
